### PR TITLE
Gitbook update

### PR DESCRIPTION
**Pull request by marcindsobczak**
* changed design of JSON RPC and CLI docs
* solved the problem of skipping tabs
* removed nesting level from JSON RPC, CLI, Metrics and Configurations
* added attribute fo JSON RPC parameter with:
  * description
  * example value
* added auto-generated request curls in JSON RPC and requests in CLI. If there are example values of parameters in method parameters attributes, curl example has example value, if not, field names are used instead
* added response description in method attribute
* added links in JSON RPC for same methods in CLI and vice versa
* added EdgeCaseHint in JSON RPC method attribute
* added sample-configuration auto-generating for:
  * mainnet.cfg
  * goerli.cfg
  * rinkeby.cfg
  * ropsten.cfg
* changed process of generating JSON RPC and CLI docs:
  * loading data first, then start generating
  * if e.g. JSON RPC method has description and/or example and corresponding CLI method doesn't have, data will be printed in both places

### CODE REVIEW

File: src/Nethermind/Nethermind.GitBook/RpcAndCliDataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+            List<Type> rpcTypes = GetRpcModules();
+            List<Type> cliTypes = GetCliModules();
+            AddRpcModulesData(rpcTypes);
+            AddCliModulesData(cliTypes);
+
+            return _modulesData;
+        }
+        
+        private List<Type> GetRpcModules()
+        {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.GitBook/RpcAndCliDataProvider.cs:44]: does it skip Baseline module and other plugins?

File: src/Nethermind/Nethermind.GitBook/RpcAndCliDataProvider.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                
+                _modulesData.Add(rpcModuleName, methods);
+            }
+        }
+                
+        private void AddCliModulesData(List<Type> cliTypes)
+        {
+            foreach (Type cliType in cliTypes)
+            {
+                MethodInfo[] moduleMethods = cliType.GetMethods().Where(method => method.Name != "GetType" && method.Name != "ToString" && method.Name != "Equals" && method.Name != "GetHashCode").ToArray();
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.GitBook/RpcAndCliDataProvider.cs:115]: not critical here but later you retrieve CliFunction / CliProperty attributes so you could use these attributes to filter out any methods that are not part of the module

**Review by tkstanczak — APPROVED** [review]

File: src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                    string returnTypeToWrite = _sharedContent.GetTypeToWrite(returnType, typesToDescribe);
+                    returnBuilder.AppendLine("| Returned type | Description |");
+                    returnBuilder.AppendLine("| :--- | :--- |");
+                    returnBuilder.AppendLine(@$"| `{returnTypeToWrite}` | {methodData.ResponseDescription} |");
+
+                    CreateCliContent(cliBuilder, paramBuilder, returnBuilder, objectsBuilder, cliMethodName, exampleArguments, cliInvocation, methodData);
+                }
+            }
+
+            const int notEmptyModule = 20;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs:193]: would be great to add a comment explaining where this 20 comes from or even do sth like:
const int emptyModuleLength = "this is empty module".Length;

File: src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+            }
+            if (cliBuilder.Length > notEmptyModule)
+            {
+                _sharedContent.Save(moduleName, string.Concat(docsDir, "/nethermind-utilities/cli"), cliBuilder);
+            }
+        }
+
+        private string CreateRpcExampleResponse(string methodDataExampleResponse)
+        {
+            string exampleResponseForRpc = methodDataExampleResponse.Replace("\n", "\n  ");
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs:207]: Environment.NewLine here instead of \n

**Review by tkstanczak — APPROVED** [review]

File: src/Nethermind/Nethermind.GitBook/RpcAndCliDataProvider.cs (hunk 1 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.GitBook/RpcAndCliDataProvider.cs:44]: Yes

File: src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs [hunk 5]
```diff
@@ -189,14 +189,14 @@ private void GenerateDocFileContent(string moduleName, Dictionary<string, Method
                     CreateCliContent(cliBuilder, paramBuilder, returnBuilder, objectsBuilder, cliMethodName, exampleArguments, cliInvocation, methodData);
                 }
             }
-
-            const int notEmptyModule = 20;
-
-            if (rpcBuilder.Length > notEmptyModule)
+            
+            int emptyModuleLength = "this is empty module".Length;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs:193]: this was just an example -> I am not sure how the empty module looks
(you had 20 there -> we need an explanation of what 20 is)

Comment (tkstanczak) [reply src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs:193]: loool, somehow I wrote something that is exactly 20 characters long but it is just a coincidence...

Comment (marcindsobczak) [reply src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs:193]: I know, but "this is empty module".Length is exactly 20 and fits perfectly. Empty module has only "# {module_name}" and nothing more

File: src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs (hunk 4 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.GitBook/RpcAndCliGenerator.cs:207]: Here I'm converting CLI example to RPC example and this replace operates on string with "\n" inside and just adding two more spaces on the beginning of every line to have RPC example formatted well after adding

{
  "jsonrpc":  "2.0",
  "result":

on the beginning, not sure if changing it have sense

**Review by tkstanczak — APPROVED** [review]

**Pull request merged**